Doc/library/sqlite3.rst

```diff
@@ -1868,12 +1868,16 @@ How to use placeholders to bind values in SQL queries
 
 SQL operations usually need to use values from Python variables. However,
 beware of using Python's string operations to assemble queries, as they
-are vulnerable to `SQL injection attacks`_ (see the `xkcd webcomic
-<https://xkcd.com/327/>`_ for a humorous example of what can go wrong)::
-
-   # Never do this -- insecure!
-   symbol = 'RHAT'
-   cur.execute("SELECT * FROM stocks WHERE symbol = '%s'" % symbol)
+are vulnerable to `SQL injection attacks`_. For example::
+
+   >>> # Never do this -- insecure!
+   >>> # An attacker can select all stocks with a crafted input.
+   >>> symbol = input()
```

Comment:
Perhaps explain further what's happening here (point out the importance of the single-quote, the logic operator, the semicolon and the SQL line comment). It's hard to balance this; we can easily fall in the trap of over-explaining stuff :)

Reply:
Maybe readers can find details in Wikipedia above the example. @erlend-aasland

Reply:
I don't think we should go into too much detail, but to me it would seem quite valuable to at least explicitly state the fundamental mechanism of the attack here (that without sanitization, an attacker can simply close the quote and write their own arbitrary query (or close the statement, and write another), instead of leaving the reader to figure it out all on their own from a Wikipedia article.

Reply:
I agree with what @CAM-Gerlach said. Also, perhaps the example would be easier to parse for a beginner if the logic was changed to simply `TRUE` instead of `1 = 1`.

Reply:
IMO, the mention now makes more clear _how_ the attacker injected arbitrary logic, but also elided the example of _what_ the logic actually does, which helps more concretely establish _why_ that is particularly harmful (which to be fair, most readers can probably assume, but IMO it might be worth being explicit and driving the point home).

To that end, it may be a little cliche, but maybe its worth having the example be something a little more clearly and unambiguously harmful to further emphasize the danger, e.g. something like the 'ole drop table? But maybe I'm going a bit overboard...

Reply:
You cannot execute multiple statements with `Cursor.execute`, so inputting `'; DROP TABLE stocks` would luckily not work. However, being able to alter the resulting row set is also a security concern.

BTW, I like the updated version very much. but I agree the last part ("arbitrary logic") can be improved,

Reply:
Ah, I see. In that case, `executescript` _would_ be vulnerable to that, no? Perhaps worth adding an example using it, to demonstrate a potentially even more serious attack (i.e. arbitrary SQL execution)?

Reply:
Yeah, f-strings and `executescript` is a _very_ dangerous combo, but `executescript` does not take parameters (other than the SQL script), so you'd really want to shoot yourself in the foot for that one. The docs tell you to use `execute` and `executemany` if you need to pass parameters to the query, so that's what most people will use. Perhaps it would be worth it to add such an example as well, but it would fail to illustrate the main point: we want people to use placeholders instead of formatted strings.

Reply:
For this PR, let's just focus on improving the existing example.

Reply:
https://github.com/python/cpython/blob/c3fdb0686041e72e3724e0915c5ad4272cc23ac8/Doc/library/sqlite3.rst#L1871-L1872
How about this improved version?